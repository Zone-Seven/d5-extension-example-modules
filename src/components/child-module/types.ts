import { ModuleEditProps } from '@divi/module-library';
import {
  CssAttr,
  FormatBreakpointStateAttr,
  IconFontAttr,
  IdClassesGroupAttr,
  InternalAttrs,
  LinkGroupAttr,
  ModuleMetaAttrs,
  PickedElementDecorationAttrs,
  TextGroupAttr,
} from '@divi/types';
import {
  ParentModuleAttrs,
} from "../parent-module/types";

export interface ChildModuleCssAttr extends CssAttr {
  contentContainer?: string;
  title?: string;
  content?: string;
  icon?: string;
}

export type ChildModuleCssGroupAttr = FormatBreakpointStateAttr<ChildModuleCssAttr>;

export interface ChildModuleAttrs extends InternalAttrs {
  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: ChildModuleCssGroupAttr;

  // Module
  module?: {
    meta?: ModuleMetaAttrs;
    advanced?: {
      link?: LinkGroupAttr;
      htmlAttributes?: IdClassesGroupAttr;
      text?: TextGroupAttr;
    };
    decoration?: PickedElementDecorationAttrs<
      'background' |
      'border' |
      'boxShadow' |
      'disabledOn' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'sizing' |
      'spacing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex'
    >;
  };

  // Title
  title?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    decoration?: PickedElementDecorationAttrs<'font'>;
  };

  // Content
  content?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    decoration?: PickedElementDecorationAttrs<'bodyFont'>;
  };

  // Icon
  // Icon
  icon?: {
    innerContent?: FormatBreakpointStateAttr<IconFontAttr>;
    advanced?: {
      color?: FormatBreakpointStateAttr<string>;
      size?: FormatBreakpointStateAttr<string>;
    };
  };
}

export type ChildModuleEditProps = ModuleEditProps<ChildModuleAttrs, ParentModuleAttrs>;
