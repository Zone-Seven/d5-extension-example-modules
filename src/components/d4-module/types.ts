// Divi dependencies.
import { ModuleEditProps } from '@divi/module-library';
import {
  CssAttr,
  FormatBreakpointStateAttr,
  IdClassesGroupAttr,
  InternalAttrs,
  LinkGroupAttr,
  TextGroupAttr,
  PickedElementDecorationAttrs,
  ModuleMetaAttrs,
  CssGroupAttr,
} from '@divi/types';

export interface D4ModuleCssAttr extends CssAttr {
  title?: string;
  content?: string;
}

export type D4ModuleCssGroupAttr = FormatBreakpointStateAttr<D4ModuleCssAttr>;

export interface D4ModuleAttrs extends InternalAttrs {
  // CSS options is used across multiple elements inside the module thus it deserves its own top property.
  css?: CssGroupAttr;

  module?: {
    meta?: ModuleMetaAttrs;
    advanced?: {
      link?: LinkGroupAttr;
      htmlAttributes?: IdClassesGroupAttr;
      text?: TextGroupAttr;
    };
    decoration?: PickedElementDecorationAttrs<
      'animation' |
      'background' |
      'border' |
      'boxShadow' |
      'disabledOn' |
      'filters' |
      'overflow' |
      'position' |
      'scroll' |
      'sizing' |
      'spacing' |
      'sticky' |
      'transform' |
      'transition' |
      'zIndex'
    >;
  };
  title?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    decoration?: PickedElementDecorationAttrs<'font'>;
  };
  content?: {
    innerContent?: FormatBreakpointStateAttr<string>;
    decoration?: PickedElementDecorationAttrs<'bodyFont'>;
  };
}

export type D4ModuleEditProps = ModuleEditProps<D4ModuleAttrs>;