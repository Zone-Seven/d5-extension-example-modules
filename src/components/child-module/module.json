{
  "name": "example/child-module",
  "title": "Child Module",
  "titles": "Child Modules",
  "category": "child-module",
  "attributes": {}
}
