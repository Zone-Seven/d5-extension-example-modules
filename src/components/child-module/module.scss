.et_pb_child_module {
  display: flex;
  flex-wrap: wrap;

  &_icon.et-pb-icon {
    font-size: 21px;
    padding-top: 6px;
    color: #29c4a9;
  }

  &_content_container {
    flex-grow: 1;
    flex-basis: 0%;
    padding-left: 15px;
  }

  &_title {
    font-size: 21px;
    line-height: 1em;
    font-weight: 600;
    margin-bottom: 10px;
  }

  &_content {
    font-size: 13px;
    line-height: 1.6em;
    color: #868686;
  }
}