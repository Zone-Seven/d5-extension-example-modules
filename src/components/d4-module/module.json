{
  "name": "example/d4-module",
  "title": "Divi 4 Module",
  "titles": "Divi 4 Modules",
  "category": "module",
  "d4Slug": "d4_module",
  "attributes": {}
}
