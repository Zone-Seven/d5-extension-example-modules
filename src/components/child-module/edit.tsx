// External Dependencies.
import React, { ReactElement } from 'react';

// Divi Dependencies.
import { ModuleContainer } from '@divi/module';
import { mergeAttrs } from '@divi/module-utils';
import { processFontIcon } from '@divi/icon-library';

// Local Dependencies.
import { ChildModuleEditProps } from './types';
import { defaultAttrs } from './constants';
import { defaultAttrs as parentDefaultAttrs } from '../parent-module/constants';
import { ModuleStyles } from './styles';
import { moduleClassnames } from './module-classnames';

/**
 * Static Module edit component of visual builder.
 *
 * @since ??
 *
 * @param {ChildModuleEditProps} props React component props.
 *
 * @returns {ReactElement}
 */
const ChildModuleEdit = (props: ChildModuleEditProps): ReactElement => {
  const {
    attrs,
    id,
    name,
    parentAttrs,
  } = props;

  // Merge parent module default values with parent module attributes.
  const parentModuleAttrs = mergeAttrs({
    defaultAttrs: parentDefaultAttrs,
    attrs: parentAttrs,
  });

  // Merge module default values with module attributes.
  const childModuleAttrs = mergeAttrs({
    defaultAttrs,
    attrs,
  });

  const moduleAttrs = {
    ...childModuleAttrs,
    icon: mergeAttrs({
      defaultAttrs: parentModuleAttrs?.icon,
      attrs: childModuleAttrs?.icon,
    }),
    iconColor: mergeAttrs({
      defaultAttrs: parentModuleAttrs?.iconColor,
      attrs: childModuleAttrs?.iconColor,
    }),
    iconSize: mergeAttrs({
      defaultAttrs: parentModuleAttrs?.iconSize,
      attrs: childModuleAttrs?.iconSize,
    }),
    titleFont: mergeAttrs({
      defaultAttrs: parentModuleAttrs?.titleFont,
      attrs: childModuleAttrs?.titleFont,
    }),
    text: mergeAttrs({
      defaultAttrs: parentModuleAttrs?.text,
      attrs: childModuleAttrs?.text,
    }),
    bodyFont: mergeAttrs({
      defaultAttrs: parentModuleAttrs?.bodyFont,
      attrs: childModuleAttrs?.bodyFont,
    }),
  };
  
  const icon    = moduleAttrs?.icon?.desktop?.value;
  const title   = moduleAttrs?.title?.desktop?.value;
  const content = moduleAttrs?.content?.desktop?.value;

  return (
    <ModuleContainer
      attrs={moduleAttrs}
      componentType="edit"
      id={id}
      name={name}
      stylesComponent={ModuleStyles}
      classnamesFunction={moduleClassnames}
      tag="li"
    >
      { icon && (
        <div className="child-module__icon et-pb-icon">
            {processFontIcon(icon)}
        </div>
      )}
      <div className="child-module__content-container">
        <div className="child-module__title">{title}</div>
        <div
          className="child-module__content"
          dangerouslySetInnerHTML={{ __html: content }}
        />
      </div>
    </ModuleContainer>
  );
}

export {
  ChildModuleEdit,
};
