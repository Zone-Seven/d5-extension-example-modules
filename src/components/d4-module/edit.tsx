// External Dependencies.
import React, { ReactElement } from 'react';
import classnames from 'classnames';

// Divi Dependencies.
import { ModuleContainer } from '@divi/module';
import { mergeAttrs } from '@divi/module-utils';

// Local Dependencies.
import { D4ModuleEditProps } from './types';
import { defaultAttrs } from './constants';
import { ModuleStyles } from './styles';
import { moduleClassnames } from './module-classnames';
import { ModuleScriptData } from './module-script-data';

/**
 * Divi 4 Module edit component of visual builder.
 *
 * @since ??
 *
 * @param {D4ModuleEditProps} props React component props.
 *
 * @returns {ReactElement}
 */
const D4ModuleEdit = (props: D4ModuleEditProps): ReactElement => {
  const {
    attrs,
    id,
    name,
  } = props;

  // Merge module default values with module attributes.
  const moduleAttrs = mergeAttrs({
    defaultAttrs,
    attrs,
  });

  const HeadingLevel = moduleAttrs?.titleFont?.font?.desktop?.value?.headingLevel;
  const title        = moduleAttrs?.title?.desktop?.value;
  const content      = moduleAttrs?.content?.desktop?.value;

  return (
    <ModuleContainer
      attrs={moduleAttrs}
      componentType="edit"
      id={id}
      name={name}
      stylesComponent={ModuleStyles}
      classnamesFunction={moduleClassnames}
      scriptDataComponent={ModuleScriptData}
    >
      <HeadingLevel className="d4_module_title">{title}</HeadingLevel>
      <div
        className="d4_module_content"
        dangerouslySetInnerHTML={{ __html: content }}
      />
    </ModuleContainer>
  );
}

export {
  D4ModuleEdit,
};
