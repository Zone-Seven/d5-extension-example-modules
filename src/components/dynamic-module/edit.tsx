// External Dependencies.
import React, { ReactElement, useEffect } from 'react';

// Divi Dependencies.
import { ModuleContainer } from '@divi/module';
import { mergeAttrs } from '@divi/module-utils';

// Local Dependencies.
import { DynamicModuleEditProps } from './types';
import { ModuleStyles } from './styles';
import { defaultAttrs } from './constants';
import { useGetRecentPosts } from './hooks/get-recent-posts';
import { map } from 'lodash';
import { __ } from '@wordpress/i18n';

/**
 * Static Module edit component of visual builder.
 *
 * @since ??
 *
 * @param {DynamicModuleEditProps} props React component props.
 *
 * @returns {ReactElement}
 */
const DynamicModuleEdit = (props: DynamicModuleEditProps): ReactElement => {
  const {
    attrs,
    id,
    name,
  } = props;

  // Merge module default values with module attributes.
  const moduleAttrs = mergeAttrs({
    defaultAttrs,
    attrs,
  });

  const {getPosts, posts, isLoading} = useGetRecentPosts();

  const title         = moduleAttrs?.title?.desktop?.value;
  const numberOfPosts = parseInt(moduleAttrs?.numberOfPosts?.desktop?.value);
  const TitleHeading  = moduleAttrs?.titleFont?.font?.desktop?.value?.headingLevel;
  const PostTitleHeading  = moduleAttrs?.postTitleFont?.font?.desktop?.value?.headingLevel;

  /**
   * Fetches new Portfolio Posts on parameter changes.
   */
   useEffect(() => {
    getPosts(numberOfPosts);
  }, [numberOfPosts]);

  return (
    <ModuleContainer
      attrs={moduleAttrs}
      componentType="edit"
      id={id}
      name={name}
      stylesComponent={ModuleStyles}
    >
      {
        ! isLoading && (
          <>
            {
              title && (
                <TitleHeading className="dynamic-module__title">{title}</TitleHeading>
              )
            }
            <div className="dynamic-module__post-items">
              {
                map(posts, (post) => (
                    <div className="dynamic-module__post-item">
                      <PostTitleHeading className="dynamic-module__post-item-title">
                        <a href={post?.link} onClick={() => false}>{post?.title?.rendered}</a>
                      </PostTitleHeading>
                      <div className="dynamic-module__post-item-content" dangerouslySetInnerHTML={{__html: post?.excerpt?.rendered}} />
                    </div>
                  )
                )
              }
            </div>
          </>
        )
      }
      {
        ! isLoading && posts.length < 1 && (
          <div>{__('No post found.', 'd5-module-extension-example')}</div>
        )
      }
      {
        isLoading && (
          <div>{__('Loading...', 'd5-module-extension-example')}</div>
        )
      }
    </ModuleContainer>
  );
}

export {
  DynamicModuleEdit,
};
