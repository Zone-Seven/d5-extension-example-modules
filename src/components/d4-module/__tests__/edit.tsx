import { createElement } from 'react';
import { shallow } from 'enzyme';
import {
  omit,
} from 'lodash';

import { registerAppUiStore } from '@divi/app-ui';
import { registerModule, registerModuleLibraryStore } from '@divi/module-library';
import { registerEventsStore } from '@divi/events';
import { registerKeyboardShortcutsStore } from '@divi/keyboard-shortcuts';
import { registerModalLibraryStore, registerModals } from '@divi/modal-library';
import { registerEditPostStore } from '@divi/edit-post';
import { registerSerializedPostStore } from '@divi/serialized-post';
import { registerAppPreferencesStore } from '@divi/app-preferences';
import { select } from '@divi/data';
import { registerHistoryStore } from '@divi/history';
import { registerModuleStore } from '@divi/module';
import { registerAjaxStore } from '@divi/rest';
import { registerSettingsStore } from '@divi/settings';
import { registerHelpStore } from '@divi/help';
import { registerFontsStore } from '@divi/fonts';
import { registerShortcodeModule } from '@divi/shortcode-module';

import { d4Module } from '..';
import testCases from './test-cases.json';
import { D4ModuleEditProps } from '../types';
import { D4ModuleEdit as ModuleComponent } from '../edit';

interface Expected {
  builder: string;
  frontend: string;
}

interface TestCase {
  title: string;
  args: D4ModuleEditProps;
  expected: Expected;
}

// Prepare variable to be used as Module Component that has been registered in the Module Library
// and have ModuleElements instance being passed as props + defaultAttrs merged into module attribute.
let D4ModuleEdit: typeof ModuleComponent;


beforeAll(() => {
  registerHelpStore();
  registerHistoryStore();
  registerModuleLibraryStore();
  registerShortcodeModule(true);

  registerModule(d4Module.metadata, omit(d4Module, 'metadata'));

  // Get module component from module-library store.
  D4ModuleEdit = select('divi/module-library').getModule('example/d4-module').renderers.edit;
});

beforeEach(() => {
  registerHelpStore();
  registerSettingsStore();
  registerAjaxStore();
  registerAppPreferencesStore();
  registerEventsStore();
  registerModalLibraryStore();
  registerAppUiStore();
  registerKeyboardShortcutsStore();
  registerSerializedPostStore();
  registerEditPostStore();
  registerModuleStore();

  if (! select('divi/font')) {
    registerFontsStore();
  }
});

describe.each(testCases as unknown as TestCase[])('<D4ModuleEdit />', ({ title, args }: TestCase) => {
  it(`Should match the snapshot when "${title}"`, () => {
    const component = createElement(D4ModuleEdit, args);

    // Snapshot test
    expect(shallow(component).html()).toMatchSnapshot();
  });
});

