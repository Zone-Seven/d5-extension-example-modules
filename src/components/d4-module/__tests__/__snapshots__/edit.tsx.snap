// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<D4ModuleEdit /> Should match the snapshot when "Default attributes that gets added when Divi 4 module is added." 1`] = `"<div class=\\"d4-module_d4-module-1 d4-module et_pb_bg_layout_light et_pb_module\\" data-id=\\"d4-module-1\\"><h2 class=\\"d4_module_title\\">Your Title Goes Here</h2><div class=\\"d4_module_content\\"><p>Your content goes here. Edit or remove this text inline or in the module Content settings. You can also style every aspect of this content in the module Design settings and even apply custom CSS to this text in the module Advanced settings.</p></div></div>"`;

exports[`<D4ModuleEdit /> Should match the snapshot when "Display content text style properly when the content text style is set." 1`] = `"<div class=\\"d4-module_d4-module-2 d4-module et_pb_bg_layout_light et_pb_module\\" data-id=\\"d4-module-2\\"><h2 class=\\"d4_module_title\\">Your Title Goes Here</h2><div class=\\"d4_module_content\\"><p>Your content goes here. Edit or remove this text inline or in the module Content settings. You can also style every aspect of this content in the module Design settings and even apply custom CSS to this text in the module Advanced settings.</p></div><style class=\\"et-vb-module-styles\\">.d4-module_d4-module-2 .d4_module_content {font-variant: small-caps; color: #8300E9; font-size: 18px; line-height: 1.8em;}.d4-module_d4-module-2 .d4_module_content {text-shadow: 0.08em 0.08em 0.08em #00FFC3;}</style></div>"`;

exports[`<D4ModuleEdit /> Should match the snapshot when "Display custom background, title color, content color and padding properly when the custom background, title color, content color and padding are set." 1`] = `"<div class=\\"d4-module_d4-module-3 d4-module et_pb_bg_layout_light et_pb_module\\" data-id=\\"d4-module-3\\"><h2 class=\\"d4_module_title\\">Your Title Goes Here</h2><div class=\\"d4_module_content\\"><p>Your content goes here. Edit or remove this text inline or in the module Content settings. You can also style every aspect of this content in the module Design settings and even apply custom CSS to this text in the module Advanced settings.</p></div><style class=\\"et-vb-module-styles\\">.d4-module_d4-module-3 {background-color: #8300E9;}.d4-module_d4-module-3 .d4_module_title {color: #ffffff;}.d4-module_d4-module-3 .d4_module_content {color: #ffffff;}.d4-module_d4-module-3 {padding-top: 30px; padding-right: 30px; padding-bottom: 30px; padding-left: 30px;}</style></div>"`;

exports[`<D4ModuleEdit /> Should match the snapshot when "Display custom styles properly when the custom styles are set." 1`] = `"<div class=\\"d4-module_d4-module-4 d4-module et_pb_bg_layout_light et_pb_module\\" data-id=\\"d4-module-4\\"><h3 class=\\"d4_module_title\\">Your Title Goes Here</h3><div class=\\"d4_module_content\\"><p>Your content goes here. Edit or remove this text inline or in the module Content settings. You can also style every aspect of this content in the module Design settings and even apply custom CSS to this text in the module Advanced settings.</p></div><style class=\\"et-vb-module-styles\\">.d4-module_d4-module-4 {background-color: #EDF000;}.d4-module_d4-module-4 {text-align: center;}.d4-module_d4-module-4 .d4_module_title {text-transform: uppercase;}.d4-module_d4-module-4 {padding-top: 30px; padding-right: 30px; padding-bottom: 30px; padding-left: 30px;}.d4-module_d4-module-4 {border-bottom-left-radius: 15px !important; border-bottom-right-radius: 15px !important; border-top-left-radius: 15px !important; border-top-right-radius: 15px !important;}.d4-module_d4-module-4 {box-shadow: 0px 12px 18px -6px rgba(0,0,0,0.3);}</style></div>"`;
